### PR TITLE
Initial pass through for updates to Kotlin

### PR DESCRIPTION
**Pull request by designatednerd**
- Added some stuff from swift
- updated a bunch of code examples from Java
- Threw in a whole mess of todos I don't have the ability to actually edit right now
- moved to make the style guide use soft-wrapping instead of hard wrapping

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -221,15 +209,15 @@ Indentation for blocks uses 2 spaces (not the default 4):
...
+```kotlin
 for (int i = 0; i < 10; i++) {
     Log.i(TAG, "index=" + i);
 }
 ```
 
 __GOOD:__
 
-```java
+```kotlin
```
Comment (tgblank) [inline README.md:258]: Update to Kotlin style loop:

```kotlin
for (i in 0..9) {
  Log.i(TAG, "index=" + i)
}
```

File: README.md [hunk 2]
```diff
@@ -221,15 +209,15 @@ Indentation for blocks uses 2 spaces (not the default 4):
 
 __BAD:__
 
-```java
+```kotlin
```
Comment (tgblank) [inline README.md:250]: Update to Kotlin style loop:

```kotlin
for (i in 0..9) {
    Log.i(TAG, "index=" + i)
}
```

**Comment by designatednerd** [comment]
@tgblank those two issues should now be addressed

File: README.md [hunk 3]
```diff
@@ -312,61 +315,61 @@ class MyClass {
...
 if (someTest) {
-  doSomething();
+  doSomething()
 }
-if (someTest) { doSomethingElse(); }
+if (someTest) { doSomethingElse() }
 ```
 
 
 ## Switch Statements
```
Comment (tgblank) [inline README.md]: Suggested change from Switch to When

## When Statements

Unlike switch statements in Java, when statements do not fall through. Separate cases by commas if they should be handled the same way.

Always include the `else` case.

__BAD:__

```kotlin
when (anInput) {
  1 -> doSomethingForCaseOne()
  2 -> doSomethingForCaseOneOrTwo()
  3 -> doSomethingForCaseThree()
}
```

__GOOD:__

```kotlin
when (anInput) {
  1, 2 -> doSomethingForCaseOneOrTwo()
  3 -> doSomethingForCaseThree()
  else -> println("No case satisfied")
}
```

File: README.md [hunk 4]
```diff
@@ -110,106 +109,95 @@ Written in __lowerCamelCase__. For example `setValue`.
...
+```kotlin
 XmlHttpRequest
-String url
+url: String
 findPostById
 ```
 
 ## Declarations
 
 ### Access Level Modifiers
```
Comment (tgblank) [inline README.md]: Suggested change:

## Visibility Modifiers

Only include visibility modifiers if you need something other than the default of `public`.

__BAD:__

```kotlin
public val wideOpenProperty = 1
private val myOwnPrivateProperty = "private"
```

__GOOD:__

```kotlin
val wideOpenProperty = 1
private val myOwnPrivateProperty = "private"
```

File: README.md [hunk 5]
```diff
@@ -110,106 +109,95 @@ Written in __lowerCamelCase__. For example `setValue`.
...
+```kotlin
+username: String
+twitterHandle: String
 ```
 
 ### Classes
 
-Exactly one class per source file, although inner classes are encouraged where
-scoping appropriate.
+Exactly one class per source file, although inner classes are encouraged where scoping appropriate.
```
Comment (tgblank) [inline README.md]: New section:

## Data Type Objects

Prefer `data` classes for simple data holding objects.

__BAD:__
```kotlin
class Person(val name: String) {
  override fun toString() : String {
    return "Person(name=$name)"
  }
}
```

__GOOD:__
```kotlin
data class Person(val name: String)
```

File: README.md [hunk 6]
```diff
@@ -260,31 +248,46 @@ Lines should be no longer than 100 characters long.
...
 ## Getters & Setters
 
-For external access to fields in classes, getters and setters are preferred to
-direct access of the fields. Fields should rarely be `public`.
+Unlike Java, direct access to fields in Kotlin is preferred. 
 
-However, it is encouraged to use the field directly when accessing internally
-(i.e. from inside the class). This is a performance optimization recommended
-by Google: http://developer.android.com/training/articles/perf-tips.html#GettersSetters
+If getters and setters are required, they should be declared [following Kotlin conventions](https://kotlinlang.org/docs/reference/properties.html) rather than as separate methods.
```
Comment (tgblank) [inline README.md]: Add word custom.

> If custom getters and setters...

File: README.md [hunk 7]
```diff
@@ -375,96 +378,131 @@ switch (anInput) {
...
+
+### Companion Objects
+
+** TODO: A bunch of stuff about companion objects **
+
+### Optionals
+
+Declare variables and function return types as optional with `?` where a `null` value is acceptable.
+
+Use implicitly unwrapped types declared with `!!` only for instance variables that you know will be initialized before use, such as subviews that will be set up in `onCreate` for an Activity or `onCreateView` for a Fragment.
```
Comment (tgblank) [inline README.md]: Would add or possibly replace with the following:

Prefer to use `lateinit` over optionals for properties that will be initialized in `onCreate` and `onCreateView`.

__BAD:__

```kotlin
private var mProgressBar: ProgressBar? = null
...
mProgressBar = findViewById(R.id.progress_bar) as ProgressBar
mProgressBar!!.visibility = ProgressBar.VISIBLE
```

__GOOD:__

```kotlin
private lateinit var mProgressBar: ProgressBar
...
mProgressBar = findViewById(R.id.progress_bar) as ProgressBar
mProgressBar.visibility = ProgressBar.VISIBLE
```

File: README.md [hunk 8]
```diff
@@ -375,96 +378,131 @@ switch (anInput) {
...
+var optionalSubview: UIView?
+var volume: Double?
 
-Where appropriate, XML attributes should appear in the following order:
+if let unwrappedSubview = optionalSubview {
+  if let realVolume = volume {
+    // do something with unwrappedSubview and realVolume
+  }
+}
+```
```
Comment (tgblank) [inline README.md]: New section:

## String Templates

Prefer string templates to string concatenation.

__BAD:__

```kotlin
var hello = "Hello"
var world = "World"
println(hello + " " + world)
```

__GOOD:__

```kotlin
var hello = "Hello"
var world = "World"
println("$hello $world")
```

File: README.md (hunk 7 above)
Comment (designatednerd) [reply README.md]: I think we probably need a whole section on `lateinit` - maybe you can cover that in the next PR? :trollface:

**Comment by designatednerd** [comment]
@tgblank I addressed most of your concerns but I think once this is merged, there's probably going to be plenty of opportunity for you to fork and add more stuff.

File: README.md (hunk 8 above)
Comment (designatednerd) [reply README.md]: (I somewhat lost the thread of where this is supposed to go)

Comment (tgblank) [reply README.md]: I don't see a good place for it within an existing section. Maybe a new top level section after **Semicolons**?

**Pull request merged**